Samples/MapScenarios/MapScenarios/Scenarios/WalkTime/WalkTimeViewController.swift

```diff
@@ -0,0 +1,177 @@
+            return -1
+        }
+        
+        var distance: CLLocationDistance = 0
+        
+        // The distance for rest of instructions
+        for instruction in restInstructions {
+            distance += instruction.distance
+        }
+        
+        // The distance from current location to the end of current instruction
+        if let instructionEndLocation = currentInstruction.points.last?.coordinate {
+            let userLocation = CLLocation(latitude: lastUpdateLocation.coordinate.latitude, longitude: lastUpdateLocation.coordinate.longitude);
+            let endLocation = CLLocation(latitude: instructionEndLocation.latitude, longitude: instructionEndLocation.longitude)
+            distance += userLocation.distance(from: endLocation)
+        }
+        
+        return distance
+    }
+}
+
+// MARK: - PWMapViewDelegate
+
+extension WalkTimeViewController: PWMapViewDelegate {
+    
+    func mapView(_ mapView: PWMapView!, locationManager: PWLocationManager!, didUpdateIndoorUserLocation userLocation: PWUserLocation!) {
+        lastUpdateLocation = userLocation
+    }
+    
+    func mapViewStartedSnappingLocation(toRoute mapView: PWMapView!) {
+        snappedLocation = true
+    }
+    
+    func mapViewStoppedSnappingLocation(toRoute mapView: PWMapView!) {
+        snappedLocation = false
+    }
+    
+    func mapView(_ mapView: PWMapView!, didChange instruction: PWRouteInstruction!) {
+        // Update walk time when blue dot is not acquired
+        if lastUpdateLocation == nil, let currentIndex = self.mapView.currentRoute.routeInstructions.firstIndex(of: instruction), let restInstructions = self.mapView.currentRoute.routeInstructions?[currentIndex...] {
+            var distance: Double = 0
+            for instruction in restInstructions {
+                distance += instruction.distance
+            }
+            self.walkTimeView?.updateWalkTime(distance: distance, averageSpeed: 0)
+        }
+    }
+}
+
+// MARK: - CLLocationManagerDelegate
+
+extension WalkTimeViewController: CLLocationManagerDelegate {
+    
+    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
+        switch status {
+        case .authorizedAlways, .authorizedWhenInUse:
+            // Re-register location manager
+            mapView.unregisterLocationManager()
+            let managedLocationManager = PWManagedLocationManager(buildingId: buildingIdentifier)
+            DispatchQueue.main.async { [weak self] in
+                self?.mapView.register(managedLocationManager)
+            }
+        default:
+            print("Not authorized to start PWManagedLocationManager")
+        }
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
+        if let location = locations.first {
```

Comment:
double check the delegate function's doc but i think locations.last is actually the most recent location